components/brave_shields/browser/ad_block_subscription_service_manager.cc

```diff
@@ -0,0 +1,491 @@
+
+const base::TimeDelta kListUpdateInterval = base::TimeDelta::FromDays(7);
+
+SubscriptionInfo BuildInfoFromDict(const GURL& sub_url,
+                                   const base::Value* dict) {
+  DCHECK(dict);
+  DCHECK(dict->is_dict());
+
+  SubscriptionInfo info;
+  base::JSONValueConverter<SubscriptionInfo> converter;
+  converter.Convert(*dict, &info);
+
+  info.subscription_url = sub_url;
+
+  return info;
+}
+
+const base::FilePath::CharType kSubscriptionsDir[] =
+    FILE_PATH_LITERAL("FilterListSubscriptionCache");
+
+}  // namespace
+
+AdBlockSubscriptionServiceManager::AdBlockSubscriptionServiceManager(
+    brave_component_updater::BraveComponent::Delegate* delegate,
+    AdBlockSubscriptionDownloadManager::DownloadManagerGetter
+        download_manager_getter,
+    const base::FilePath& user_data_dir)
+    : delegate_(delegate),
+      subscription_path_(user_data_dir.AppendASCII("System Profile")
+                             .AppendASCII(kSubscriptionsDir)),
+      subscriptions_(new base::DictionaryValue()) {
+  LoadSubscriptionServices();
+  std::move(download_manager_getter)
+      .Run(base::BindOnce(
+          &AdBlockSubscriptionServiceManager::OnGetDownloadManager,
+          weak_ptr_factory_.GetWeakPtr()));
+}
+
+AdBlockSubscriptionServiceManager::~AdBlockSubscriptionServiceManager() {}
+
+base::FilePath AdBlockSubscriptionServiceManager::GetSubscriptionPath(
+    const GURL& sub_url) const {
+  // Subdirectories are generated by taking the SHA256 hash of the list URL
+  // spec, then base64 encoding that hash. This generates paths that are:
+  //     - deterministic
+  //     - unique
+  //     - constant length
+  //     - path-safe
+  //     - not too long (exactly 45 characters)
+  const std::string hash = crypto::SHA256HashString(sub_url.spec());
+
+  std::string pathsafe_hash;
+  base::Base64UrlEncode(hash, base::Base64UrlEncodePolicy::INCLUDE_PADDING,
+                        &pathsafe_hash);
+
+  return subscription_path_.AppendASCII(pathsafe_hash);
+}
+
+GURL AdBlockSubscriptionServiceManager::GetListTextFileUrl(
+    const GURL sub_url) const {
+  base::FilePath cached_list_path = GetSubscriptionPath(sub_url).AppendASCII(
+      brave_shields::kCustomSubscriptionListText);
+
+  const GURL file_url = net::FilePathToFileURL(cached_list_path);
+
+  return file_url;
+}
+
+void AdBlockSubscriptionServiceManager::OnUpdateTimer(
+    const GURL& sub_url,
+    bool from_ui,
+    component_updater::TimerUpdateScheduler::OnFinishedCallback on_finished) {
+  ready_.Post(FROM_HERE,
+              base::BindOnce(&AdBlockSubscriptionServiceManager::StartDownload,
+                             weak_ptr_factory_.GetWeakPtr(), sub_url, from_ui));
+
+  std::move(on_finished).Run();
+}
+
+void AdBlockSubscriptionServiceManager::StartDownload(const GURL& sub_url,
+                                                      bool from_ui) {
+  DCHECK(ready_.is_signaled());
+  download_manager_->StartDownload(sub_url, from_ui);
```

Comment:
you need a null check here because the download manager is tied to the lifetime of the system profile, but the `AdBlockSubscriptionServiceManager` lives as long as browser process

Reply:
added the check